### ISSUE TITLE
Controlling a parameter using MIDI notes?

### ISSUE THREAD

**Issue by pdesaulniers** [DISTRHO/DPF#134]
Here is my use case:

I've got a "Pattern Index" parameter, which is an integer. It can be set by pressing switches in the UI. It allows selecting one of many user-editable graphs, which are stored as states (just like in Wolf Shaper). 

Here is a crude mockup:

![2019-04-05-15:27:18](https://user-images.githubusercontent.com/23723605/55651694-73815c00-57b7-11e9-9d16-3dfb59c4cf33.png)
![2019-04-05-15:27:28](https://user-images.githubusercontent.com/23723605/55651710-81cf7800-57b7-11e9-86f6-b13a299c9246.png)

However, I would also like to be able to select "Pattern 1" by sending C4 to the plugin, "Pattern 2" by sending C#4, etc.

Would it be possible to implement this feature in a DPF plugin? If not, could DPF be extended to support this use case?

I could send the selected pattern to the UI using an output parameter, but I don't think output parameters are necessarily saved by the host. Also, I think output parameters are readonly in the UI (unless I'm wrong?), so I wouldn't be able to change the selected pattern by clicking on a switch.

**Comment by falkTX** [comment]
You cannot a plugin modify its own (input) parameters on the DSP side.
Doing this would break DSSI and LV2 support, and gets very messy to support anyway.

Having MIDI input modifying some internal state of the plugin is completely out of question for DPF, as that design choice is questionable to begin with (the host is the one that should do midi-learn, not the plugin, but sadly not everyone can see this).

The only way you have in DPF to do what you want is to enable the "full state" macro.
See https://github.com/DISTRHO/DPF/blob/master/distrho/DistrhoPlugin.hpp#L838
With this macro on you break DSSI compatibility, but yeah...
What it does is make the host (on save) call `getState` of the plugin for all keys, without having to pass a message from ui to dsp.
You can use this to save your custom state, while having an output parameter to report things to the UI.

So in short:
1. enable `DISTRHO_PLUGIN_WANT_FULL_STATE` macro
2. add a new state key that will have the selected pattern
3. Implement `getState` to save that value (and being restored on project load by the host)
4. Add an output parameter to report the current selected pattern to the UI
5. on the UI, use `setState` to modify the state key for the selected pattern

on the dsp side you will need to handle `setState` and modify the output parameter
on the UI, the state value is ignored, and you use the output parameter instead.

I hope that makes sense

**Comment by pdesaulniers** [comment]
Ah yes, that makes sense! Thanks!

**Comment by falkTX** [comment]
This is something I added for zynaddsubfx, has not been needed for other plugins so far.
I simply forgot to document it in the info header together with the rest.
